specs/message-passing/message-passing-unified-specs.md

```diff
@@ -0,0 +1,221 @@
+    participant GatewayA as Gateway (Chain A)
+    participant HookA as Hook (Chain A)
+    participant Relayer as Off-Chain Relayer
+    participant GatewayB as Gateway (Chain B)
+    participant Recipient as IMessageRecipient
+
+    User->>GatewayA: sendMessage(recipient, payload, extraData, hookData)
+    GatewayA->>HookA: hookData
+    GatewayA->>GatewayA: MessageSent (messageId, sender, recipient, payload)
+    GatewayA-->>Relayer: Off-chain bridging flow
+
+    Note over GatewayB: using deliverMessage as example method, destination Gateway is out of scope
+    Relayer-->>GatewayB: deliverMessage(metadata, message)
+    GatewayB->>GatewayB: runVerifications()
+    GatewayB->>Recipient: receiveMessage(messageId, sender, payload)
+
+```
+
+```mermaid
+graph TD
+ Relayer
+ 
+ subgraph Origin Chain 
+	 GatewayE[Gateway]
+	 Sender --"sendMessage(recipient, payload, extraData, hookData)"-->GatewayE
+	 GatewayE --"run HookData()"-->HookContract
+ end
+ 
+
+ subgraph Destination Chain
+	 _GatewayR_[Gateway]
+	 Recipient
+	 	 
+	 _GatewayR_ --"_run verifications(☁️)_"--> _GatewayR_
+	 _GatewayR_ --"receiveMessage(messageId, sender, payload)"--> Recipient
+ end
+ Relayer --"_deliverMessage(messageId, sender, payload)_"--> _GatewayR_
+
+```
+
+## Rationale and Discussions
+
+This design is an incremental evolution based on ERC-7786 and other related standards, aligning progressively with what is considered best practice.
+
+**What stays the same**
+
+- **Unchanged core API:** send and receive messages behave exactly as in ERC-7786. Adapters would remain the same way.
+- **Safety / Liveness expectations**: All guarantees listed in ERC-7786 are inherited unchanged.
+- **Event-driven tracking**: an `outboxId` is still emitted on the origin chain, and a `messageId` is provided by the relaying/deliver step on the destination chain.
+
+**What is added**
+
+- **Hooks (optional):** A  `HookData` struct lets a project plug in fee payment, logging, batched calls, callbacks or any other external logic without bloating the base call.
+- **Future‑proof addressing:** Binary interoperable addresses ([ERC-7930](https://github.com/ethereum/ERCs/pull/1002)); 7786 plans to adopt the same format, so this is forward‑compatible.
+- **Clear separation of concerns**: `extraData` tweaks behavior inside the gateway, while `HookData` runs code outside the gateway.
+
+**Hooks vs. Attributes**
+
+ERC-7786 introduces attributes (key/value blobs interpreted by the gateway). Hooks fill the same niche but with two advantages:
+
+1. Arbitrary Logic: A hook is external contract logic that the message sender wants to execute. Hooks remove the limitation of relying solely on the selectors implemented by the Gateway, enabling developers to execute arbitrary functions the Gateway didn’t originally anticipate.
+2. No bloat for minimal gateways: A bare-bones gateway doesn’t need to parse or store attribute blobs. If a project wants richer behavior, it can deploy its own hook contract.
```

Comment:
Note that a bare-bones gateway in current ERC-7786 doesn't need to handle attributes, it can return `false` in `supportsAttributes` and have `require(attributes.length == 0)` within `sendMessage`.